src/Host/Client/Impl/Session/RSession.cs

```diff
@@ -262,34 +266,31 @@ internal sealed class RSession : IRSession, IRCallbacks {
 
             await _initializationTcs.Task;
             initializationCts.Dispose();
+            _stopHostLock.EnqueueReset();
             lockToken.Set();
         }
 
-        public IRSessionSwitchBrokerTransaction StartSwitchingBroker() {
-            _disposeToken.ThrowIfDisposed();
-            return new BrokerTransaction(this);
-        }
+        public IRSessionSwitchBrokerTransaction StartSwitchingBroker() => !_disposeToken.IsDisposed ? new BrokerTransaction(this) : null;
 
-        public IRSessionReconnectTransaction StartReconnecting() {
-            _disposeToken.ThrowIfDisposed();
-            return new BrokerTransaction(this);
-        }
+        public IRSessionReconnectTransaction StartReconnecting() => !_disposeToken.IsDisposed ? new BrokerTransaction(this) : null;
 
         public async Task StopHostAsync() {
             _disposeToken.ThrowIfDisposed();
```

Comment:
In case of a switch to the background thread, if caller is on the main thread, continuation will be immediately executed in a thread pool. If it is on the background thread already, then no awaiting will happen at all.

It is different in case of UI thread though. We need to add `cancellationToken` to the `SwitchToMainThreadAsync`.

---
In reply to: [83911726](https://github.com/Microsoft/RTVS/pull/2529#discussion_r83911726) [](ancestors = 83911726)